FOLIO holds configuration and update

ihardy opened the pull request:
Add holds config copied from VoyagerRestful.ini. Noticed holds links not showing up for Folio, @mis306lu noted an empty `$holdsConfig` and @demiankatz suggested adding in config in this request to get started on this issue.

TODO:
- [x] Implement comments, or remove comments from extraHoldFields
- [x] Implement pickup location, or remove comments from extraHoldFields
- [x] Fix other problems causing errors when placing holds
- [x] Resolve [VUFIND-1390](https://vufind.org/jira/browse/VUFIND-1390) when merging

demiankatz commented:
Thanks for getting the ball rolling, @ihardy! I've removed a lot of what you copied-and-pasted to eliminate Voyager-specific details that are unlikely to apply here. What's left is the bare minimum; some of the rest we might copy back if we add associated features (like implementing `getDefaultPickupLocation()`).

I notice that the FOLIO driver does not currently implement support for the "comments" or "pickUpLocation" settings in the extraHoldFields option. I've added TODO items to either implement that functionality or else remove the settings from the file.

I'm not sure if FOLIO supports user comments on holds; if not, removing the option is the obvious thing to do.

Regarding pickup location, I'm wondering if the hard-coded "Hold Shelf" in the driver (https://github.com/vufind-org/vufind/blob/master/module/VuFind/src/VuFind/ILS/Driver/Folio.php#L822) needs to be updated to use the incoming pickup location data; in any case, that value doesn't appear to be getting used at all at present.

Could you and/or @mis306lu take another look and let me know your thoughts on these finishing touches? Hopefully we can get this merged without too much additional work.

ihardy commented:
Thanks @demiankatz I don't believe FOLIO supports user comments on holds. I do think the hard coded hold-shelf can be replaced, the loan schema in FOLIO takes the uuid of a service point as "pickupServicePointId".

demiankatz commented:
Thanks, @ihardy, I've removed the comments field for now. It sounds like wiring up the pickup location in place of the hard-coded value should be fairly straightforward, as long as the getPickupLocations() method is already returning the appropriate UUIDs.

As I suggested in Slack, it probably also makes sense to fix the hold problems @mis306lu ran into today. Please let me know if you need my help to take a stab at any of this.

demiankatz commented on module/VuFind/src/VuFind/ILS/Driver/Folio.php:
```diff
@@ -808,33 +808,32 @@ public function placeHold($holdDetails)
         $response = $this->makeRequest(
             'POST',
-            '/request-storage/requests',
+            '/circulation/requests/instances',
```
As mentioned on Slack, this looks like it is going to place a title-level, rather than item-level hold; VuFind expects item-level holds by default, and will request title-level holds by passing 'title' as the 'level' key of the $holdDetails array. (See https://github.com/vufind-org/vufind/blob/master/module/VuFind/src/VuFind/ILS/Driver/VoyagerRestful.php#L1936 for the Voyager driver's example). If we can support both types of hold here, that would be great, but I don't think we want to rely purely on title-level holds, since that can lead to unexpected results (especially with multi-volume sets). Let me know if you need any clarification!

mis306lu replied:
I think FOLIO supports title-level holds (https://s3.amazonaws.com/foliodocs/api/mod-circulation/p/circulation.html#circulation_requests_instances_post)??   However it requires the instance UUID ...which we were not planning to pass into VuFind.  So at this stage in the process I have the bib HRID and the item UUID which I think I want to use to place the request.
I'm not sure how to make this all configurable / flexible.  

Also, if VuFind is going to alllow item based holds, I think it does need the hold type (page or hold).   In the holdingsils/standard.phtml file where the hold links are constructed it knows the item status...so I'm wondering if it can be appended to the url (e.g. &available=$holding['availability']  .....or something like that)?

I changed my code to be item based, added fulfilmentPreference and requestType and it works.  I'm just not sure how to pull all the options (title level/item level) together.  

I'll paste my code below.  

 $requestBody = [
            'itemId' => $holdDetails['item_id'],
            'requesterId' => $holdDetails['patron']['id'],
            'requestDate' => date('c'),
            'requestExpirationDate' => date_format($requiredBy, 'Y-m-d'),
            'pickupServicePointId' => $holdDetails['pickUpLocation'],
	    'fulfilmentPreference' => 'Hold Shelf',
	    'requestType' => 'Page'  //hardcoded for now
        ];
        $response = $this->makeRequest(
            'POST',
            '/circulation/requests',
            json_encode($requestBody)
        );

thanks!

demiankatz replied:
@mis306lu, I think we just need to use getInstanceByBibId() to look up the instance ID at the top of placeHold() -- this can be used to translate HRIDs into instance IDs in your case. Perhaps we want to create a new getInstanceIdByBibId() method so that we can short-circuit the API call in situations where we already have an instance ID.

Regarding the page vs. hold thing, the Voyager driver receives a 'holdtype' value from VuFind (see https://github.com/vufind-org/vufind/blob/c23c92c734aaae8b4b288f948e21ef9d19c7d5bc/module/VuFind/src/VuFind/ILS/Driver/VoyagerRestful.php#L1934) which it uses to place the correct type of request. I can't remember off the top of my head exactly where holdtype is coming from, but we should be able to leverage that mechanism in FOLIO as well. I'd start by var_dumping the incoming array to see what details are in there... if 'holdtype' is missing by default, let me know and I'll go digging into the remaining details.

mis306lu replied:
The incoming array to placeHold looks like this:

array(5) { ["requiredBy"]=> string(10) "06-07-2020" ["pickUpLocation"]=> string(36) "c4c90014-c8c9-4ade-8f24-b5e313319f4b" ["id"]=> string(2) "43" ["item_id"]=> string(36) "91569542-0172-4081-abc8-ad50e8cf2ff8" ["patron"]=> array(7) { ["id"]=> string(36) "1dd51453-1165-45b4-a5d5-cc948483f61e" ["username"]=> string(9) "876579559" ["cat_username"]=> string(9) "5551212" ["cat_password"]=> NULL ["firstname"]=> string(8) "Jane" ["lastname"]=> string(10) "Doe" ["email"]=> string(18) "an email address" } }

ihardy replied:
On the subject of instanceIds vs HRIDs, I got a little turned around (see slack) and this should be using `/circulation/requests` instead of `/circulation/requests/instance` OR `/requests-storage/requests`. Like Demian pointed out this can be an item level request instead of instance like I've done here. 

There's no `auto` hold type in FOLIO to fall back on, so I think we'll need to add back in that $status variable which I took out to try to pick the correct hold type to place.

demiankatz replied:
@mis306lu, I think the way to get holdtype to show up in the incoming array is to add it as a value to the HMACKeys setting in FOLIO.ini, and to include a 'holdtype' key in the array returned by getHolding() (at the same level as the is_holdable key). The idea is that, at the same time you're looking up holding status, you can determine what type of request would be appropriate, and then that value gets passed back around to the placeHold() method. The code is not crystal clear and I'm reviewing it in a hurry, so I apologize if I lead you astray -- but I think it's worth a quick experiment. If it doesn't work, please let me know and I'll dig deeper.

mis306lu replied:
I'll give it a try.  thank you!

mis306lu replied:
I ended up with this:
added 'status' to the HMACKeys...and then I was able to add this back into the placeHold function:

	    'requestType' => $holdDetails['status'] == 'Available' ? 'Page' : 'Recall',

Checked out items AND Paged items work with both request type Recall and request type Hold.  I need to look into FOLIO to see what the difference is.  Do you think this approach is okay?    thank you!!

ihardy replied:
Thanks @mis306lu and @demiankatz I updated the PR catch up with this conversation I think, let me know if I've misinterpreted your changes. Also updated to use the `/circulation/requests` interface to place a request by item id.

mis306lu replied:
this looks good to me.  thank you!

demiankatz commented on module/VuFind/src/VuFind/ILS/Driver/Folio.php:
```diff
@@ -808,33 +808,35 @@ public function placeHold($holdDetails)
             ];
         } else {
-            throw new ILSException($response->getBody());
+            $json = json_decode($response->getBody());
```
It would probably be a good idea to account for the possibility that the response is not JSON here, in case of a more generalized server failure.

ihardy replied:
Thanks, forgive my basic php, but I'm thinking something like this
```
            try {
                $json = json_decode($response->getBody());
                $result = [
                    'success' => false,
                    'status' => $json->errors[0]->message
                ];
            } catch (Exception $e) {
                throw new ILSException($response->getBody());
            }
```
Would it be better still to not to assume we can get a response and just throw `$e->getMessage()`?

demiankatz replied:
I think what you propose makes sense. I suppose we could concatenate both $e->getMessage and the body together for the most comprehensive possible exception message, but without a real-world example, it's hard to know whether that would be better or worse. We can start here and revise as experience informs us to. :-)

demiankatz commented on module/VuFind/src/VuFind/ILS/Driver/Folio.php:
```diff
@@ -808,33 +808,35 @@ public function placeHold($holdDetails)
-            ],
             'itemId' => $holdDetails['item_id'],
+            'requestType' => $holdDetails['status'] == 'Available'
+                ? 'Page' : 'Recall',
```
Do we always want to place a "Recall" on an unavailable item, or does this need to be configurable? (I'm assuming that "Hold" is different behavior from "Recall" -- at least, in Voyager, there's the difference between notifying the current holder of the item that they must return it early, vs. simply waiting until they return it and notifying the user with the hold).

mis306lu replied:
I wondered about this also.  I'll research and see what I can find.

demiankatz replied:
Thanks!

mis306lu replied:
This [JIRA ](https://issues.folio.org/browse/UXPROD-1558 ) points to this [Google doc](https://docs.google.com/spreadsheets/d/1jaID-HGft3q4YzJq6Ycy2ejWByAIvou3Zyyw4ko87YI/edit#gid=0) which describes each hold type in detail.

demiankatz replied:
Very helpful; that does suggest that there's likely a need for this to be more configurable, since different institutions are going to have different request type preferences. Let me know if you'd like me to help with the design of that.

demiankatz commented on module/VuFind/src/VuFind/ILS/Driver/Folio.php:
```diff
@@ -808,33 +808,35 @@ public function placeHold($holdDetails)
-                'lastName' => $holdDetails['patron']['lastname'] ?? ''
-            ],
             'itemId' => $holdDetails['item_id'],
+            'requestType' => $holdDetails['status'] == 'Available'
```
Is the 'Available' status hard-coded into FOLIO, or might we need to configure this to recognize different statuses under different configurations of the system?

mis306lu replied:
From what I've seen, "Available" is hard-coded into FOLIO.

demiankatz replied:
I guess the related question is whether there are any other statuses that might influence the type of request we would want to place. If there's any possibility of variation, it would be easy enough to create a configurable map in the class and/or .ini file to correlate status strings with request types (and have an overall default to use if nothing in the map is matched).

ihardy commented:
I think I've addressed the comments above in de9abd5. I also added cancelHolds in 27d1ab1. (If this should be a separate pr, lmk and I'll revert 27d1ab1 and make a separate request after this one is merged). A few notes on the hold cancellation:

* Modified getMyHolds to only show requests where the status is open. Possible statuses are available here: https://github.com/folio-org/mod-circulation/blob/master/ramls/request.json#L40
* Also modified array returned by getMyHolds to include the requestId, itemId, and to derive the request type instead of hard coding.

Question about  `cancelHolds`:
* I think the easiest way to cancel a hold is to get the request, modify it, and PUT it back. Its an extra call, but information like pickupservicepoint isn't in the returned value of getMyHolds. Should getting the request information be a separate private function, or is it OK as I've done it all in the cancelHolds function?

Finally, I noticed a potential problem. We're not setting limits on requests to the FOLIO APIs. Say a user has 25 hold requests in FOLIO. I think the default limit is 10 so we'll only be seeing a portion of a user's holds from getMyHolds. This isn't limited to this function, so maybe we can discuss it on slack sometime. Might make sense to implement some kind of reusable helper function to page through results to make sure the number of items returned == the number found (unless I'm overlooking something that already exists).

demiankatz commented:
Thanks, @ihardy! I'll review this in more detail tomorrow when I'm back to work. Regarding limits/pagination, that definitely sounds like something in need of fixing. I did implement pagination for the course reserves functionality (https://github.com/vufind-org/vufind/blob/master/module/VuFind/src/VuFind/ILS/Driver/Folio.php#L916) so perhaps that code can be further generalized.

demiankatz commented:
Okay, @ihardy, I had time to look at this more closely. It's shaping up very nicely. I think the cancelHolds function is fine as-is; we can always refactor it to smaller methods in future if there is a need.

So I think this leaves two questions:

1.) Do we want to try to resolve the limit/pagination issue here, or should we get this merged as-is and then address that more holistically in a separate PR? (If I had to choose, I'd probably favor the latter, but since I'm not doing most of the work, my opinion is not the most important).

2.) Should getMyHolds() be limiting the type of requests being retrieved to Holds and Recalls? I want to be sure we don't display other types of requests out of their appropriate context (e.g. what if we implement Storage Retrieval Requests in the future?)

Let me know what you think, and hopefully we can merge soon (perhaps after another round of testing from @mis306lu, if she has time).

ihardy commented:
Thanks again for the review. 

1.) I think it makes sense to address the pagination thing in a separate PR since we'll need to apply it to a few functions that aren't really holds related.

2.) The only request types we should be getting back from /circulation/requests or /request-storage/requests are "Page", "Hold", and "Recall" (see: https://github.com/folio-org/mod-circulation-storage/blob/master/ramls/request.json#L14). So I think getMyHolds is OK how it was originally written in that regard. I suppose if a new request type is added later we could be surprised.

demiankatz commented:
Thanks, @ihardy; that all makes sense, so I think this is ready to merge. I'll wait a few days in case @mis306lu has any further input, however. I also added a cross-link with [VUFIND-1390](https://vufind.org/jira/browse/VUFIND-1390) so we remember to resolve that ticket when this gets merged.

mis306lu commented:
Looks great to me.  Tested holds and cancel holds.  thank you for working on this!

demiankatz commented:
Thanks, @mis306lu -- merging now!
